tiatoolbox/tools/registration/wsi_registration.py

```diff
@@ -536,6 +531,356 @@ def estimate_affine_transform(
+                fixed_tissue.shape[1],
+            ],
+            [
+                int(np.ceil(fixed_tissue.shape[0] / 2)),
+                fixed_tissue.shape[0],
+                0,
+                int(np.floor(fixed_tissue.shape[1] / 2)),
+            ],
+            [
+                int(np.ceil(fixed_tissue.shape[0] / 2)),
+                fixed_tissue.shape[0],
+                int(np.ceil(fixed_tissue.shape[1] / 2)),
+                fixed_tissue.shape[1],
+            ],
+        ]
+        fixed_matched_points, moving_matched_points, quality = [], [], []
+        for _index, bounding_box in enumerate(blocks_bounding_box):
+            fixed_block = fixed_tissue[
+                bounding_box[0] : bounding_box[1], bounding_box[2] : bounding_box[3], :
+            ]
+            moving_block = moving_tissue[
+                bounding_box[0] : bounding_box[1], bounding_box[2] : bounding_box[3], :
+            ]
+            features = self.extract_features(fixed_block, moving_block)
+            (
+                fixed_block_matched_points,
+                moving_block_matched_points,
+                block_quality,
+            ) = self.feature_mapping(features)
+            fixed_matched_points.append(
+                fixed_block_matched_points + [bounding_box[2], bounding_box[0]]
+            )
+            moving_matched_points.append(
+                moving_block_matched_points + [bounding_box[2], bounding_box[0]]
+            )
+            quality.append(block_quality)
+        fixed_matched_points, moving_matched_points, quality = (
+            np.concatenate(fixed_matched_points),
+            np.concatenate(moving_matched_points),
+            np.concatenate(quality),
+        )
+        (
+            fixed_matched_points,
+            moving_matched_points,
+            quality,
+        ) = self.filtering_matching_points(
+            fixed_mask,
+            moving_mask,
+            fixed_matched_points,
+            moving_matched_points,
+            quality,
+        )
+        block_transform = DFBRegister.estimate_affine_transform(
+            fixed_matched_points, moving_matched_points
+        )
+
+        translation_ = np.array(
+            [
+                [1, 0, -tissue_top_left_coord[1]],
+                [0, 1, -tissue_top_left_coord[0]],
+                [0, 0, 1],
+            ]
+        )
+        translation = np.array(
+            [
+                [1, 0, tissue_top_left_coord[1]],
+                [0, 1, tissue_top_left_coord[0]],
+                [0, 0, 1],
+            ]
+        )
+        tissue_transform = np.vstack([tissue_transform, np.array([0, 0, 1])])
+        block_transform = np.vstack([block_transform, np.array([0, 0, 1])])
+        return np.matmul(
+            np.matmul(
+                np.matmul(np.matmul(translation, block_transform), tissue_transform),
+                translation_,
+            ),
+            transform_initializer,
+        )
```

Comment:
You might want to break this up a bit for readability. It's a bit hard to parse like this.

Reply:
Not that you can also use `@` for matrix multiplication. e.g. `X @ t`

Reply:
https://github.com/TissueImageAnalytics/tiatoolbox/blob/3225a5835708a89cd5999385b406aa3d1a663fe8/tiatoolbox/tools/registration/wsi_registration.py#L1038-L1040